Add a warning that ImplicitlyUnwrappedOptional is deprecated

rudkx opened the pull request:
Per SE-0054, implicitly unwrapped optional is not a distinct type in the type system, but rather just the notion that certain Optionals (denoted by the sigil "!" rather than "?") can be implicitly unwrapped.

This is a first step in the direction of implementing this notion by emitting a warning if the type is spelled out.
    
Later on this will be reworked a bit by having "ImplicitlyUnwrappedOptional" resolve to the Optional type for versions < 5, with the standard library type ImplicitlyUnwrappedOptional completely removed (so versions >= 5 will error out on references).

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
Thoughts on this?

The warnings are currently emitted regardless of swift-version, which is presumably the wrong thing to do since they mention a specific version and emitting it for earlier versions could break people building with -warnings-as-errors.

slavapestov commented on lib/Sema/TypeCheckPattern.cpp:
```diff
@@ -816,6 +816,21 @@ bool TypeChecker::typeCheckPattern(Pattern *P, DeclContext *dc,
+    // Temporarily add a deprecation warning and when possible a fixit
+    // to replace ImplicitlyUnwrappedOptional<T> with T! so that we
+    // can eliminate the type ImplicitlyUnwrappedOptional in time.
+    auto *IUODecl = Context.getImplicitlyUnwrappedOptionalDecl();
```
I think this should be moved to resolveType() in TypeCheckType.cpp. It's not just patterns where we want to prohibit explicitly spelling out IUO, but everywhere else too, eg

```
func foo(_: ImplicitlyUnwrappedOptional<Int>) {}
func foo() -> ImplicitlyUnwrappedOptional<Int> {}
func foo<T : Sequence>(_: T) where T.Element ==  ImplicitlyUnwrappedOptional<Int> {}
```

And so on.

This will also cover the extension case.

slavapestov replied:
Also

```
typealias T = ImplicitlyUnwrappedOptional<Int>
```

Actually right now even this is allowed, and it should not be

```
typealias T = Int!
```

rudkx replied:
Yeah I had it implemented somewhere under there originally and then moved it to emit distinct diagnostics for the extension case, but now I realize doing so means we're emitting diagnostics for '!' as well.

rudkx replied:
I don't think we want to ban ``typealias T = Int!`` if that's what you're suggesting. There's no great reason people shouldn't be able to do that and then declare 

```
func f(_ x: T) -> T { ... }
```

We do need to make sure these cannot appear anywhere that a directly written ``!`` appears.

rudkx replied:
BTW, I view the fact that we can inject IUOs in places where they shouldn't be allowed as completely orthogonal and it's probably the last thing I'll actually get to unless it just happens to fall out.

slavapestov replied:
I see what you're saying, but I don't think there's much of benefit from defining such an alias. Also right now we can do stuff like

```
typealias T = Int!
func f(_: Array<T>) {}
```

Which should be banned, too.

Now we could allow the type alias but ban the array of it from being formed by looking through underlying types when we do our IUO checks, but that is more complex than just banning IUO in certain positions at the TypeRepr level.

I think the intent of the original IUO evolution proposal was that IUO is literally only allowed on parameters, returns, and properties, only directly and not via alias.

jrose-apple replied:
I agree with Slava. The intent of SE-0054 was to make IUO not part of the type system at all.

DougGregor replied:
Piling on agreement with @slavapestov and @jrose-apple that SE-0054 implies that `!` should not be part of the type system at all. It's only allowed when directly attributable to a parameter, result type, or property.

jrose-apple replied:
Right, a typealias is an alias at the Type level, not the TypeRepr level.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2977,6 +2977,15 @@ WARNING(swift3_ignore_specialized_enum_element_call,none,
         "cannot specialize enum case; ignoring generic argument, "
         "which will be rejected in future version of Swift", ())
 
+WARNING(bound_implicitly_unwrapped_optional_spelling_deprecated,none,
```
These should be ERRORs in -swift-version 5 mode.

rudkx replied:
Sure.

slavapestov reviewed (commented):
Nice!

> Note that extensions of ImplicitlyUnwrappedOptional (and Optional for
that matter) do not appear to work at the moment anyway, so banning them
should not have any impact.

I thought extensions of Optional did work. I've seen people define them before. A simple example I tried appears to work:

```
(swift) extension Optional { func hello() { print("I'm an Optional of \(Wrapped.self)") } }
(swift) let x: Int? = 5
// x : Int? = Optional(5)
(swift) x.hello()
I'm an Optional of Int
```

> The warnings are currently emitted regardless of swift-version, which is presumably the wrong thing to do since they mention a specific version and emitting it for earlier versions could break people building with -warnings-as-errors.

We don't promise source compatibility with -warnings-as-errors, for example the stdlib introduces deprecations that are not conditional on the -swift-version.

rudkx commented:
I'm going to rework this tomorrow and force-push a new commit once it's revised.

Commit:
Add a warning that ImplicitlyUnwrappedOptional is deprecated in 4+.

Per SE-0054, implicitly unwrapped optional is not a distinct type in the
type system, but rather just the notion that certain Optionals (denoted
by the sigil "!" rather than "?") can be implicitly unwrapped.

This is a first step in the direction of implementing this notion by
emitting a warning if the type is spelled out.

slavapestov commented on lib/Sema/TypeChecker.h:
```diff
@@ -536,6 +536,9 @@ enum TypeResolutionFlags : unsigned {
   TR_SubscriptParameters = 0x2000000,
+
+  /// Is it okay to resolve an IUO sigil ("!") here?
+  TR_AllowIUO = 0x4000000,
```
I'm assuming plumbing this through to be checked in more places and ripping out the old diagnostic is still pending?

rudkx replied:
Yes. I'll probably look at that next although I have a couple other changes I could finish up and push as well.

slavapestov reviewed (commented):
Looks good, thanks! Were you going to eventually use TR_AllowIUO to diagnose invalid uses of the sugared form as well?

rudkx commented:
@swift-ci Please smoke test

DougGregor commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -1180,6 +1180,32 @@ resolveTopLevelIdentTypeComponent(TypeChecker &TC, DeclContext *DC,
+  if (currentDecl == IUODecl && TC.Context.isSwiftVersionAtLeast(4)) {
+    if (current->getNominalOrBoundGenericNominal() &&
+        options.contains(TR_AllowIUO)) {
+      auto *genericTyR = cast<GenericIdentTypeRepr>(comp);
```
I'm not sure we've hit the point where `current->getNominalOrBoundGenericNominal()` implies that `cast<GenericIdentTypeRepr>(comp)` will succeed. I recommend gating this on `isa< GenericIdentTypeRepr>(comp)` instead.

rudkx replied:
Ah, okay.

DougGregor reviewed (commented):
Looking much cleaner with this update.

DougGregor commented:
Actually... we should deprecate in Swift 3 as well.

rudkx commented:
@DougGregor In which case I can just remove the version check at this point, yes?

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please smoke test Linux platform

rudkx commented:
Okay I think this is done for real unless someone has more feedback.

davezarzycki commented:
Hi everybody. Would you all be open to breaking `TR_AllowIUO` into `TR_FunctionResult` and `TR_TopLevel`? I'm asking because I have added `TR_FunctionResult` on a side branch, and it seems silly to have two highly overlapping type resolution flags.

rudkx commented:
@davezarzycki Yes, I think that's a good idea. I just noticed that my change is still missing cases it should be catching (I think) because it's not treating `TR_AllowIUO` as context that gets stripped out in `withoutContext`, so I am going to look into fixing that, and splitting this into context flags as part of that fix makes sense.

Do you want to open a PR that just adds your `TR_FunctionResult` enumeration value?

davezarzycki commented:
Hi @rudkx! I don't need a PR, but if that is helpful, then by all means. As an aside, it seems to like the type resolution flags shouldn't embed policy, and therefore `TR_AllowIUO` should be broken into policy-free flags that might be useful in other contexts too. :-)
